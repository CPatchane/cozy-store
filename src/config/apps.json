{
  "notRemovableApps": ["drive"],
  "notDisplayedApps": ["collect", "home", "settings", "store"]
}
