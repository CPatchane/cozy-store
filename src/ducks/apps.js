/* eslint-env browser */

import { combineReducers } from 'redux'

import {
  UnavailableStackException,
  UnauthorizedStackException
} from '../lib/exceptions'

const FETCH_MY_APPS = 'FETCH_MY_APPS'
const FETCH_MY_APPS_SUCCESS = 'FETCH_MY_APPS_SUCCESS'
const FETCH_MY_APPS_FAILURE = 'FETCH_MY_APPS_FAILURE'

const list = (state = [], action) => {
  switch (action.type) {
    case FETCH_MY_APPS_SUCCESS:
      return action.myApps
    default:
      return state
  }
}

const isFetching = (state = false, action) => {
  switch (action.type) {
    case FETCH_MY_APPS:
      return true
    case FETCH_MY_APPS_SUCCESS:
    case FETCH_MY_APPS_FAILURE:
      return false
    default:
      return state
  }
}

export const error = (state = null, action) => {
  switch (action.type) {
    case FETCH_MY_APPS_FAILURE:
      return action.error
    default:
      return state
  }
}

export const myAppsReducers = combineReducers({
  list,
  error,
  isFetching
})

// Since apps ressource is not available from cozy-client-js
const STACK_DOMAIN = '//' + document.querySelector('[role=application]').dataset.cozyDomain
const STACK_TOKEN = document.querySelector('[role=application]').dataset.cozyToken

// the option credentials:include tells fetch to include the cookies in the
// request even for cross-origin requests
function fetchOptions () {
  return {
    credentials: 'include',
    headers: {
      Authorization: `Bearer ${STACK_TOKEN}`
    }
  }
}

async function getIcon (url) {
  const res = await fetch(`${STACK_DOMAIN}${url}`, fetchOptions())
  // res.text if SVG, otherwise res.blob  (mainly for safari support)
  const resClone = res.clone() // res must be cloned to be used twice
  const blob = await res.blob()
  const text = await resClone.text()

  try {
    return 'data:image/svg+xml;base64,' + btoa(text)
  } catch (e) { // eslint-disable-line
    return URL.createObjectURL(blob)
  }
}

const PLATFORM_APPS = ['settings', 'drive', 'collect']
export function fetchApps (cozyUrl, token) {
  return (dispatch, getState) => {
    dispatch({type: FETCH_MY_APPS})
    return fetch(`${STACK_DOMAIN}/apps/`, fetchOptions())
    .then(res => {
      if (res.status === 401) {
        dispatch({type: FETCH_MY_APPS_FAILURE, error: new UnauthorizedStackException()})
        throw new UnauthorizedStackException()
      }
      return res.json()
    })
    .then(json => {
      const data = json.data
      Promise.all(data.map(app => {
        return getIcon(app.links.icon)
        .then(iconData => {
          return Object.assign({}, app.attributes, {
            _id: app.id,
            icon: iconData,
            removable: !PLATFORM_APPS.includes(app.attributes.slug)
          })
        })
      }))
      .then(myApps => {
        dispatch({type: FETCH_MY_APPS_SUCCESS, myApps})
        return myApps
      })
    })
    .catch(e => {
      dispatch({type: FETCH_MY_APPS_FAILURE, error: e})
      throw new UnavailableStackException()
    })
  }
}
