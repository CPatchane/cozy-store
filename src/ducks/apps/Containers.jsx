import { connect } from 'react-redux'

import {
  fetchApps,
  fetchInstalledApps,
  fetchLatestApp,
  uninstallApp,
  getInstalledApps,
  getRegistryApps,
  installAppFromRegistry,
  installApp
} from './index'

import DiscoverComponent from './components/Discover'
import HiddenInstallerViewComponent from './components/HiddenInstallerView'
import { translate } from 'cozy-ui/react/I18n'
import MyApplicationsComponent from './components/MyApplications'

const mapStateToProps = (state, ownProps) => ({
  apps: getRegistryApps(state),
  installedApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  actionError: state.apps.actionError,
  fetchError: state.apps.fetchError
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchApps: () => dispatch(fetchApps(ownProps.lang)),
  fetchInstalledApps: () => dispatch(fetchInstalledApps(ownProps.lang)),
  installApp: (appSlug, appType, channel) =>
    dispatch(installAppFromRegistry(appSlug, appType, channel)),
  uninstallApp: (appSlug, appType) => dispatch(uninstallApp(appSlug, appType)),
  updateApp: (appSlug, appType, channel) =>
    dispatch(installAppFromRegistry(appSlug, appType, channel, true)),
  fetchLatestApp: (slug, channel) => dispatch(fetchLatestApp(slug, channel)),
  // for the hidden installer only
  installUsingInstaller: (appSlug, appType, source, isUpdate) =>
    dispatch(installApp(appSlug, appType, source, isUpdate)).catch(() => {
      dispatch({
        type: 'SEND_LOG_FAILURE',
        alert: {
          message: `HiddenInstallerView.${
            isUpdate ? 'update' : 'install'
          }_error`,
          messageData: { slug: appSlug },
          level: 'error'
        }
      })
    })
})

// translate last to pass the lang property to fetchApps()
export const Discover = translate()(connect(mapStateToProps, mapDispatchToProps)(
  DiscoverComponent
))

export const HiddenInstallerView = translate()(connect(mapStateToProps, mapDispatchToProps)(
  HiddenInstallerViewComponent
))

export const MyApplications = translate()(connect(mapStateToProps, mapDispatchToProps)(
  MyApplicationsComponent
))
