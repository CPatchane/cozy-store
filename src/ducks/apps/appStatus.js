export const getCurrentStatusLabel = app => {
  const { installed, maintenance, availableVersion } = app
  if (availableVersion) return 'update'
  if (maintenance) return 'maintenance'
  if (installed) return 'installed'
  return null
}

export const hasPendingUpdate = app => {
  if (app.availableVersion) return true
  return false
}

export const isUnderMaintenance = app => {
  if (!app.availableVersion && app.maintenance) return true
  return false
}

/* installed here means no actions needed */
export const isInstalledAndNothingToReport = app => {
  if (!app.availableVersion && !app.maintenance && app.installed) return true
  return false
}
