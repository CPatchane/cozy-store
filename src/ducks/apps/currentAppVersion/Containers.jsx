import { connect } from 'react-redux'

import { fetchLastAppVersion } from './index'

import InstallModalComponent from '../components/InstallModal'
import ApplicationPageComponent from '../components/ApplicationPage'

const mapStateToProps = (state, ownProps) => ({
  currentAppVersion: state.currentAppVersion.version,
  isVersionFetching: state.currentAppVersion.isFetching,
  versionError: state.currentAppVersion.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchLastAppVersion: (appSlug, channel) => dispatch(fetchLastAppVersion(appSlug, channel))
})

export const InstallModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(InstallModalComponent)

export const ApplicationPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(ApplicationPageComponent)
