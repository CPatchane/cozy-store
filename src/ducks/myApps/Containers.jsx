import { connect } from 'react-redux'

import { fetchApps, fetchMyApps, installApp, uninstallApp, getInstalledApps } from './index'

import MyApplicationsComponent from './components/MyApplications'
import DiscoverComponent from './components/Discover'

const mapStateToProps = (state, ownProps) => ({
  apps: state.apps.list,
  myApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  error: state.apps.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchApps: () => dispatch(fetchApps()),
  fetchMyApps: () => dispatch(fetchMyApps()),
  installApp: (appSlug, channel) => dispatch(installApp(appSlug, channel)),
  uninstallApp: (appSlug) => dispatch(uninstallApp(appSlug))
})

export const MyApplications = connect(
  mapStateToProps,
  mapDispatchToProps
)(MyApplicationsComponent)

export const Discover = connect(
  mapStateToProps,
  mapDispatchToProps
)(DiscoverComponent)
