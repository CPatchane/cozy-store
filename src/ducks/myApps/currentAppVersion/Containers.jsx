import { connect } from 'react-redux'

import { fetchLastAppVersion } from './index'

import InstallModalComponent from '../components/InstallModal'

const mapStateToProps = (state, ownProps) => ({
  currentAppVersion: state.currentAppVersion.version,
  isVersionFetching: state.currentAppVersion.isFetching,
  versionError: state.currentAppVersion.error
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchLastAppVersion: (appSlug, channel) => dispatch(fetchLastAppVersion(appSlug, channel))
})

export const InstallModal = connect(
  mapStateToProps,
  mapDispatchToProps
)(InstallModalComponent)
