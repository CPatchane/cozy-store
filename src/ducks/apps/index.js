/* eslint-env browser */
/* global cozy */

import { combineReducers } from 'redux'
import { currentAppVersionReducers } from './currentAppVersion'
import config from 'config/apps'

import {
  NotUninstallableAppException
} from '../../lib/exceptions'

const APP_STATE = {
  READY: 'ready',
  INSTALLING: 'installing',
  ERRORED: 'errored'
}

const FETCH_APPS = 'FETCH_APPS'
const FETCH_APPS_SUCCESS = 'FETCH_APPS_SUCCESS'
const FETCH_APPS_FAILURE = 'FETCH_APPS_FAILURE'

const FETCH_REGISTRY_APPS_SUCCESS = 'FETCH_REGISTRY_APPS_SUCCESS'

const UNINSTALL_APP_SUCCESS = 'UNINSTALL_APP_SUCCESS'
const UNINSTALL_APP_FAILURE = 'UNINSTALL_APP_FAILURE'

const INSTALL_APP = 'INSTALL_APP'
const INSTALL_APP_SUCCESS = 'INSTALL_APP_SUCCESS'
const INSTALL_APP_FAILURE = 'INSTALL_APP_FAILURE'

export const list = (state = [], action) => {
  switch (action.type) {
    case FETCH_REGISTRY_APPS_SUCCESS:
      return _sortAlphabetically(_consolidateApps(state, action.apps), 'slug')
    case FETCH_APPS_SUCCESS:
      return _sortAlphabetically(_consolidateApps(state, action.apps), 'slug')
    case UNINSTALL_APP_SUCCESS:
    case INSTALL_APP_SUCCESS:
      return _sortAlphabetically(action.apps, 'slug')
    default:
      return state
  }
}

export const isFetching = (state = false, action) => {
  switch (action.type) {
    case FETCH_APPS:
      return true
    case FETCH_APPS_SUCCESS:
    case FETCH_APPS_FAILURE:
      return false
    default:
      return state
  }
}

export const isInstalling = (state = false, action) => {
  switch (action.type) {
    case INSTALL_APP:
      return true
    case INSTALL_APP_SUCCESS:
    case INSTALL_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const actionError = (state = null, action) => {
  switch (action.type) {
    case UNINSTALL_APP_FAILURE:
    case INSTALL_APP_FAILURE:
      return action.error
    case UNINSTALL_APP_SUCCESS:
    case INSTALL_APP_SUCCESS:
      return null
    default:
      return state
  }
}

export const fetchError = (state = null, action) => {
  switch (action.type) {
    case FETCH_APPS_FAILURE:
      return action.error
    case FETCH_APPS_SUCCESS:
      return null
    default:
      return state
  }
}

export const appsReducers = combineReducers({
  list,
  actionError,
  fetchError,
  isFetching,
  isInstalling,
  currentAppVersion: currentAppVersionReducers
})

export function getInstalledApps (state) {
  return state.apps.list.filter(app => app.installed)
}

export function getRegistryApps (state) {
  // display only apps with stable versions for now
  return state.apps.list.filter(app => app.isInRegistry).filter(app => (Array.isArray(app.versions.stable) && !!app.versions.stable))
}

function _sortAlphabetically (array, property) {
  return array.sort((a, b) => a[property] > b[property])
}

async function _getIcon (url) {
  if (!url) return ''
  let icon
  try {
    icon = await cozy.client.fetchJSON('GET', url)
  } catch (e) {
    return ''
  }

  try {
    return 'data:image/svg+xml;base64,' + btoa(icon)
  } catch (e) { // eslint-disable-line
    try {
      return URL.createObjectURL(icon)
    } catch (e) {
      return ''
    }
  }
}

function _consolidateApps (stateApps, newAppsInfos) {
  const apps = new Map()
  stateApps.forEach(app => apps.set(app.slug, app))
  newAppsInfos.forEach(app => {
    const appFromState = apps.get(app.slug)
    if (appFromState) {
      apps.set(app.slug, Object.assign({}, appFromState, app))
    } else {
      apps.set(app.slug, app)
    }
  })
  return Array.from(apps.values()).filter(app => app)
}

export function fetchInstalledApps () {
  return (dispatch, getState) => {
    dispatch({type: FETCH_APPS})
    return cozy.client.fetchJSON('GET', '/apps/')
    .then(installedApps => {
      installedApps = installedApps.filter(app => !config.notDisplayedApps.includes(app.attributes.slug))
      Promise.all(installedApps.map(app => {
        // FIXME waiting name and description is locales object everywhere
        const appDesc = typeof app.attributes.description === 'string'
          ? { en: app.attributes.description } : app.attributes.description
        const appName = typeof app.attributes.name === 'string'
          ? { en: app.attributes.name } : app.attributes.name
        return _getIcon(app.links.icon)
        .then(iconData => {
          return Object.assign({}, app.attributes, {
            _id: app.id,
            icon: iconData,
            name: appName,
            description: appDesc,
            installed: true,
            related: app.links.related,
            uninstallable: !config.notRemovableApps.includes(app.attributes.slug)
          })
        })
      }))
      .then(apps => {
        return dispatch({type: FETCH_APPS_SUCCESS, apps})
      })
    })
    .catch(e => {
      dispatch({type: FETCH_APPS_FAILURE, error: e})
      throw e
    })
  }
}

export function fetchRegistryApps () {
  return (dispatch, getState) => {
    dispatch({type: FETCH_APPS})
    return cozy.client.fetchJSON('GET', '/registry?filter[type]=webapp')
    .then(response => {
      const apps = response.data
      .filter(app => !config.notDisplayedApps.includes(app.name))
      .filter(app => app.versions.dev && app.versions.dev.length) // only apps with versions available
      return Promise.all(apps.map(app => {
        return _getIcon(`/registry/${app.slug}/icon`)
        .then(iconData => {
          return Object.assign({}, app, {
            icon: iconData,
            installed: false,
            uninstallable: true,
            isInRegistry: true
          })
        })
      }))
      .then(apps => {
        return dispatch({type: FETCH_REGISTRY_APPS_SUCCESS, apps})
      })
    })
    .catch(e => {
      dispatch({type: FETCH_APPS_FAILURE, error: e})
      throw e
    })
  }
}

export function fetchApps () {
  return (dispatch, getState) => {
    dispatch(fetchRegistryApps())
    .then(() => dispatch(fetchInstalledApps()))
  }
}

export function uninstallApp (slug) {
  return (dispatch, getState) => {
    if (config.notRemovableApps.includes(slug) || config.notDisplayedApps.includes(slug)) {
      const error = new NotUninstallableAppException()
      dispatch({ type: UNINSTALL_APP_FAILURE, error })
      throw error
    }
    return cozy.client.fetchJSON('DELETE', `/apps/${slug}`)
    .then(() => {
      // remove the app from the state apps list
      const apps = getState().apps.list.map(app => {
        if (app.slug === slug) app.installed = false
        return app
      })
      dispatch({type: UNINSTALL_APP_SUCCESS, apps})
      return dispatch({
        type: 'SEND_LOG_SUCCESS',
        alert: {
          message: 'app_modal.uninstall.message.success',
          level: 'success'
        }
      })
    })
    .catch(e => {
      dispatch({type: UNINSTALL_APP_FAILURE, error: e})
      throw e
    })
  }
}

export function installApp (slug, source, isUpdate = false) {
  return (dispatch, getState) => {
    dispatch({type: INSTALL_APP})
    const verb = isUpdate ? 'PUT' : 'POST'
    return cozy.client.fetchJSON(verb, `/apps/${slug}?Source=${encodeURIComponent(source)}`)
    .then(resp => waitForAppReady(resp))
    .then(appData => {
      return _getIcon(appData.links.icon)
      .then(iconData => {
        return Object.assign({}, appData.attributes, {
          _id: appData.id,
          icon: iconData,
          installed: true,
          uninstallable: !config.notRemovableApps.includes(appData.attributes.slug)
        })
      })
      .then(app => {
        // add the installed app to the state apps list
        const apps = getState().apps.list.map(a => {
          if (a.slug === slug) {
            return Object.assign({}, a, app, {installed: true})
          }
          return a
        })
        dispatch({type: INSTALL_APP_SUCCESS, apps})
        return dispatch({
          type: 'SEND_LOG_SUCCESS',
          alert: {
            message: `app_modal.install.message.${isUpdate ? 'update' : 'install'}_success`,
            level: 'success'
          }
        })
      })
    })
    .catch(e => {
      dispatch({type: INSTALL_APP_FAILURE, error: e})
      throw e
    })
  }
}

export function installAppFromRegistry (slug, channel = 'stable') {
  return (dispatch, getState) => {
    const source = `registry://${slug}/${channel}`
    return dispatch(installApp(slug, source, false))
  }
}

// monitor the status of the app and resolve when the app is ready
function waitForAppReady (app, timeout = 20 * 1000) {
  if (app.attributes.state === APP_STATE.READY) return app
  return new Promise((resolve, reject) => {
    let idTimeout
    let idInterval

    idTimeout = setTimeout(() => {
      clearInterval(idInterval)
      resolve(app)
    }, timeout)

    idInterval = setInterval(() => {
      cozy.client.fetchJSON('GET', `/apps/${app.attributes.slug}`)
        .then(app => {
          if (app.attributes.state === APP_STATE.ERRORED) {
            if (idTimeout) {
              clearTimeout(idTimeout)
            }

            clearInterval(idInterval)
            reject(new Error('Error when installing the application'))
          }

          if (app.attributes.state === APP_STATE.READY) {
            if (idTimeout) {
              clearTimeout(idTimeout)
            }

            clearInterval(idInterval)
            resolve(app)
          }
        })
        .catch(error => {
          if (error.status === 404) return // keep waiting
          if (idTimeout) {
            clearTimeout(idTimeout)
          }

          clearInterval(idInterval)
          reject(error)
        })
    }, 1000)
  })
}
