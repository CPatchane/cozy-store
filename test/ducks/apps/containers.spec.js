'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import { MyApplications, Discover } from 'ducks/apps/Containers'

const mockStore = configureStore()

describe('Apps Containers (connected components):', () => {
  const initialState = {
    apps: {
      list: [],
      isFetching: false,
      isInstalling: false,
      actionError: null,
      fetchError: null
    }
  }

  it('MyApplications should be rendered correctly at initial store', () => {
    const store = mockStore(initialState)
    const component = shallow(<MyApplications store={store} />)
      .dive()
      .getElement()
    expect(component).toMatchSnapshot()
  })

  it('Discover should be rendered correctly at initial store', () => {
    const store = mockStore(initialState)
    const component = shallow(<Discover store={store} />)
      .dive()
      .getElement()
    expect(component).toMatchSnapshot()
  })
})
