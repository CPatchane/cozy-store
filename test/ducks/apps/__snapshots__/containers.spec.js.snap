// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps Containers (connected components): Discover should be rendered correctly at initial store 1`] = `
<Wrapper
  actionError={null}
  apps={Array []}
  f={undefined}
  fetchError={null}
  fetchLatestApp={[Function]}
  installApp={[Function]}
  installedApps={Array []}
  isAppFetching={undefined}
  isFetching={false}
  isInstalling={false}
  isUninstalling={undefined}
  lang={undefined}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  t={undefined}
  uninstallApp={[Function]}
  updateApp={[Function]}
/>
`;

exports[`Apps Containers (connected components): MyApplications should be rendered correctly at initial store 1`] = `
<Wrapper
  actionError={null}
  apps={Array []}
  f={undefined}
  fetchError={null}
  fetchLatestApp={[Function]}
  installApp={[Function]}
  installedApps={Array []}
  isAppFetching={undefined}
  isFetching={false}
  isInstalling={false}
  isUninstalling={undefined}
  lang={undefined}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  t={undefined}
  uninstallApp={[Function]}
  updateApp={[Function]}
/>
`;
